### PR TITLE
Support CreateNode and DeleteNode in REST API

### PR DESCRIPTION
**Pull request by cnxdeveloper**
## Issue
By default, in restful API mode, Flower only supports clients running as anonymous. This makes it difficult to manage tasks and results for each client.
## Proposal
-  Update the 2 POST API requests to create and delete a node.
- Update the client code to support calling the API to create a node and delete a node.
- Update the code and README in ```flower/examples/mt-pytorch``` to run and test with restful mode.
### Checklist

- [ ] Implement proposed change
- [ ] Write tests
- [ ] Update [documentation](https://flower.dev/docs/writing-documentation.html)
- [ ] Update [changelog](https://github.com/adap/flower/blob/main/doc/source/changelog.rst)
- [ ] Make CI checks pass
- [ ] Ping maintainers on [Slack](https://flower.dev/join-slack/) (channel `#contributions`)

### CODE REVIEW

**Comment by cnxdeveloper** [comment]
> Thanks for the well-scoped PR @cnxdeveloper! It looks pretty clean and easy to review, I will try to do a full review by the end of this week. In the meantime, could you already add an entry to the changelog in the `Unreleased` section?

Hi @danieljanes,
I really like the Flower framework. It's easy to use and scale up. Thank you for creating such an awesome repo!
I haven't updated the changelog in the ```Unreleased``` section yet. I'll do that later today.

File: examples/mt-pytorch/client.py [hunk 1]
```diff
@@ -35,8 +36,16 @@ def evaluate(self, parameters, config):
...
+        client=FlowerClient(),
+        rest=True,
+        transport="rest",
+    )
+else:
+    fl.client.start_numpy_client(
+        server_address="0.0.0.0:9092",
+        client=FlowerClient(),
+        transport="grpc-rere",
+    )
```
Comment (danieljanes) [inline examples/mt-pytorch/client.py:51]: `rest=True` is deprecated and will be removed. How about just adding comments to make it easier to switch between gRPC-rere and REST?
```suggestion
    fl.client.start_numpy_client(
        server_address="0.0.0.0:9092",  # "0.0.0.0:9093" for REST
        client=FlowerClient(),
        transport="grpc-rere",  # "rest" for REST
    )
```

File: doc/source/ref-changelog.md [hunk 2]
```diff
@@ -1,7 +1,7 @@
 # Changelog
 
 ## Unreleased
-
+-  **Update REST API to support create and delete nodes** ([#2283](https://github.com/adap/flower/pull/2283))
```
Comment (danieljanes) [inline doc/source/ref-changelog.md:4]:
```suggestion

-  **Update REST API to support create and delete nodes** ([#2283](https://github.com/adap/flower/pull/2283))

```

File: examples/mt-pytorch/README.md [hunk 3]
```diff
@@ -49,3 +49,26 @@ Terminal 2+3: start two clients
 ```bash
 python client.py
 ```
+
+## Run Restful API mode
```
Comment (danieljanes) [inline examples/mt-pytorch/README.md:53]:
```suggestion
## Run with Driver API (REST transport layer)
```

File: examples/mt-pytorch/README.md [hunk 4]
```diff
@@ -49,3 +49,26 @@ Terminal 2+3: start two clients
...
+
+
+Terminal 1: start Flower server restful mode
```
Comment (danieljanes) [inline examples/mt-pytorch/README.md:56]:
```suggestion
Terminal 1: start Flower server and enable REST transport layer
```

File: examples/mt-pytorch/README.md [hunk 5]
```diff
@@ -49,3 +49,26 @@ Terminal 2+3: start two clients
...
+
+```bash
+flower-server --rest
+```
+
+Terminal 2: start driver
```
Comment (danieljanes) [inline examples/mt-pytorch/README.md:62]:
```suggestion
Terminal 2: start Driver script
```

File: examples/mt-pytorch/README.md [hunk 6]
```diff
@@ -49,3 +49,26 @@ Terminal 2+3: start two clients
...
+
+```bash
+python driver
```
Comment (danieljanes) [inline examples/mt-pytorch/README.md:65]:
```suggestion
python driver.py
```

File: examples/mt-pytorch/README.md [hunk 7]
```diff
@@ -49,3 +49,26 @@ Terminal 2+3: start two clients
...
+```
+
+Open file ```client.py``` set variable ```restful_mode``` is ```True```
```
Comment (danieljanes) [inline examples/mt-pytorch/README.md:68]:
```suggestion
Open file `client.py` adjust `server_address` and `transport`.
```

File: src/py/flwr/client/rest_client/connection.py [hunk 8]
```diff
@@ -114,17 +121,99 @@ def http_request_response(
...
+            return False
+        
+        # Deserialize ProtoBuf from bytes
+        create_node_response_proto = CreateNodeResponse()
+        create_node_response_proto.ParseFromString(res.content)
+   
+        node_store[KEY_NODE] = create_node_response_proto.node
+        return True
+    
+    def delete_node() -> bool:
```
Comment (danieljanes) [inline src/py/flwr/client/rest_client/connection.py:171]: Missing docstring

File: src/py/flwr/server/fleet/rest_rere/rest_api.py [hunk 9]
```diff
@@ -33,6 +38,60 @@
...
+    
+    # Get state from app
+    state: State = app.state.STATE_FACTORY.state()
+    
+    # Handle message
+    delete_node_response = message_handler.delete_node(
+        request=delete_node_request_proto,
+        state=state
+    )
+    # Return Serialize Protobuf
```
Comment (danieljanes) [inline src/py/flwr/server/fleet/rest_rere/rest_api.py:86]:
```suggestion

    # Return serialized ProtoBuf
```

File: src/py/flwr/server/fleet/rest_rere/rest_api.py [hunk 10]
```diff
@@ -33,6 +38,60 @@
...
+    )
+
+
+async def delete_node(request: Request) -> Response:
+    """Delete Node Id"""
+    _check_headers(request.headers)
+
+    # Get the request body as raw bytes
+    delete_node_request_bytes: bytes = await request.body()
+    # Deserialize ProtoBuf
```
Comment (danieljanes) [inline src/py/flwr/server/fleet/rest_rere/rest_api.py:74]:
```suggestion

    # Deserialize ProtoBuf
```

File: src/py/flwr/server/fleet/rest_rere/rest_api.py [hunk 11]
```diff
@@ -33,6 +38,60 @@
...
+    
+    # Get state from app
+    state: State = app.state.STATE_FACTORY.state()
+    
+    # Handle message
+    new_node_data = message_handler.create_node(
+        request=create_node_request_proto, 
+        state=state
+    )
+    # Return serialized ProtoBuf
```
Comment (danieljanes) [inline src/py/flwr/server/fleet/rest_rere/rest_api.py:59]:
```suggestion

    # Return serialized ProtoBuf
```

File: src/py/flwr/server/fleet/rest_rere/rest_api.py [hunk 12]
```diff
@@ -33,6 +38,60 @@
     sys.exit(MISSING_EXTRA_REST)
 
 
+async def create_node(request: Request) -> Response:
+    """Create Node"""
+    _check_headers(request.headers)
+
+    # parse base 64 to string
+    create_node_request_bytes: bytes = await request.body()
+    # Deserialize ProtoBuf
```
Comment (danieljanes) [inline src/py/flwr/server/fleet/rest_rere/rest_api.py:47]:
```suggestion

    # Deserialize ProtoBuf
```

File: src/py/flwr/server/fleet/rest_rere/rest_api.py [hunk 13]
```diff
@@ -33,6 +38,60 @@
     sys.exit(MISSING_EXTRA_REST)
 
 
+async def create_node(request: Request) -> Response:
+    """Create Node"""
+    _check_headers(request.headers)
+
+    # parse base 64 to string
```
Comment (danieljanes) [inline src/py/flwr/server/fleet/rest_rere/rest_api.py:45]:
```suggestion
    # Parse base 64 to string
```

File: src/py/flwr/client/rest_client/connection.py [hunk 14]
```diff
@@ -179,6 +268,12 @@ def receive() -> Optional[TaskIns]:
 
     def send(task_res: TaskRes) -> None:
         """Send task result back to server."""
+        # get Node
```
Comment (danieljanes) [inline src/py/flwr/client/rest_client/connection.py:271]:
```suggestion
        # Get Node
```

File: src/py/flwr/client/rest_client/connection.py [hunk 15]
```diff
@@ -114,17 +121,99 @@ def http_request_response(
...
         # Serialize ProtoBuf to bytes
-        pull_task_ins_req_proto = PullTaskInsRequest(
-            node=Node(node_id=0, anonymous=True),
-        )
+        if node_store[KEY_NODE] is None:
+            log(ERROR, "Node instance missing")
+            return None
+        node: Node = cast(Node, node_store[KEY_NODE])
+        
+        pull_task_ins_req_proto = PullTaskInsRequest( node=node)
```
Comment (danieljanes) [inline src/py/flwr/client/rest_client/connection.py:216]:
```suggestion
        pull_task_ins_req_proto = PullTaskInsRequest(node=node)
```

**Review by danieljanes — CHANGES_REQUESTED** [review]
The PR is in a good shape 👍 I added some comments to align the coding style with the rest of the codebase

File: src/py/flwr/client/rest_client/connection.py (hunk 8 above)
Comment (cnxdeveloper) [reply src/py/flwr/client/rest_client/connection.py:171]: I have already updated

File: src/py/flwr/client/rest_client/connection.py [hunk 16]
```diff
@@ -114,17 +121,100 @@ def http_request_response(
...
         # Serialize ProtoBuf to bytes
-        pull_task_ins_req_proto = PullTaskInsRequest(
-            node=Node(node_id=0, anonymous=True),
-        )
+        if node_store[KEY_NODE] is None:
+            log(ERROR, "Node instance missing")
+            return None
+        node: Node = cast(Node, node_store[KEY_NODE])
+        
+        pull_task_ins_req_proto = PullTaskInsRequest(node=node)
```
Comment (danieljanes) [inline src/py/flwr/client/rest_client/connection.py:217]:
```suggestion
        # Request instructions (task) from server
        pull_task_ins_req_proto = PullTaskInsRequest(node=node)
```

File: src/py/flwr/server/fleet/rest_rere/rest_api.py [hunk 17]
```diff
@@ -33,6 +38,64 @@
     sys.exit(MISSING_EXTRA_REST)
 
 
+async def create_node(request: Request) -> Response:
+    """Create Node"""
+    _check_headers(request.headers)
+
+    # Parse base 64 to string
```
Comment (danieljanes) [inline src/py/flwr/server/fleet/rest_rere/rest_api.py:45]:
```suggestion
    # Get the request body as raw bytes
```

File: src/py/flwr/server/fleet/rest_rere/rest_api.py [hunk 18]
```diff
@@ -33,6 +38,64 @@
...
+        state=state
+    )
+
+    # Return serialized ProtoBuf
+    new_node_data_bytes = new_node_data.SerializeToString()
+    return Response(
+        status_code=200,
+        content=new_node_data_bytes,
+        headers={"Content-Type": "application/protobuf"},
+    )
```
Comment (danieljanes) [inline src/py/flwr/server/fleet/rest_rere/rest_api.py:67]:
```suggestion
    create_node_response_proto = message_handler.create_node(
        request=create_node_request_proto,
        state=state
    )

    # Return serialized ProtoBuf
    create_node_response_bytes = create_node_response_proto.SerializeToString()
    return Response(
        status_code=200,
        content=create_node_response_bytes,
        headers={"Content-Type": "application/protobuf"},
    )
```

File: src/py/flwr/server/fleet/rest_rere/rest_api.py [hunk 19]
```diff
@@ -33,6 +38,64 @@
...
+        state=state
+    )
+
+    # Return serialized ProtoBuf
+    delete_node_response_bytes = delete_node_response.SerializeToString()
+    return Response(
+        status_code=200,
+        content=delete_node_response_bytes,
+        headers={"Content-Type": "application/protobuf"},
+    )
```
Comment (danieljanes) [inline src/py/flwr/server/fleet/rest_rere/rest_api.py:96]:
```suggestion
    # Handle message
    delete_node_response_proto = message_handler.delete_node(
        request=delete_node_request_proto,
        state=state
    )

    # Return serialized ProtoBuf
    delete_node_response_bytes = delete_node_response_proto.SerializeToString()
    return Response(
        status_code=200,
        content=delete_node_response_bytes,
        headers={"Content-Type": "application/protobuf"},
    )
```

**Review by danieljanes — CHANGES_REQUESTED** [review]
Thanks for the fast updates, @cnxdeveloper! I added a few comments to improve consistency and I think that we can merge the PR once they are resolved.

**Comment by danieljanes** [comment]
@cnxdeveloper there's a formatting error: https://github.com/adap/flower/actions/runs/6120808796/job/16614584948?pr=2283

This can be fixed by running `./dev/format.sh`

**Comment by cnxdeveloper** [comment]
> @cnxdeveloper there's a formatting error: https://github.com/adap/flower/actions/runs/6120808796/job/16614584948?pr=2283
> 
> This can be fixed by running `./dev/format.sh`

Hi @danieljanes I fixed code format by script  `./dev/format.sh`

File: src/py/flwr/client/rest_client/connection.py [hunk 20]
```diff
@@ -114,17 +120,98 @@ def http_request_response(
...
+                "Accept": "application/protobuf",
+                "Content-Type": "application/protobuf",
+            },
+            data=create_node_req_bytes,
+            verify=verify,
+        )
+
+        # Check status code and headers
+        if res.status_code != 200:
+            return False
```
Comment (danieljanes) [inline src/py/flwr/client/rest_client/connection.py:147]:
```suggestion
            return
```

**Review by danieljanes — CHANGES_REQUESTED** [review]
This should resolve the CI error

File: src/py/flwr/client/rest_client/connection.py [hunk 21]
```diff
@@ -179,6 +266,12 @@ def receive() -> Optional[TaskIns]:
 
     def send(task_res: TaskRes) -> None:
         """Send task result back to server."""
+        # Get Node
+        if node_store[KEY_NODE] is None:
+            log(ERROR, "Node instance missing")
+            return None
```
Comment (danieljanes) [inline src/py/flwr/client/rest_client/connection.py:272]:
```suggestion
            return
```

File: src/py/flwr/client/rest_client/connection.py [hunk 22]
```diff
@@ -114,17 +120,98 @@ def http_request_response(
...
+                "[Node] POST /%s: header `Content-Type` has wrong value",
+                PATH_PULL_TASK_INS,
+            )
+            return
+
+        # Deserialize ProtoBuf from bytes
+        create_node_response_proto = CreateNodeResponse()
+        create_node_response_proto.ParseFromString(res.content)
+
+        node_store[KEY_NODE] = create_node_response_proto.node
```
Comment (danieljanes) [inline src/py/flwr/client/rest_client/connection.py:167]:
```suggestion
        node_store[KEY_NODE] = create_node_response_proto.node  # pylint: disable=no-member
```

**Review by danieljanes — CHANGES_REQUESTED** [review]

**Comment by danieljanes** [comment]
I fixed a few linter warnings, could you reformat once more @cnxdeveloper?

**Comment by danieljanes** [comment]
Almost there! Only formatting issues in two markdown files left: https://github.com/adap/flower/actions/runs/6138546228/job/16655523326?pr=2283#step:6:252

**Comment by cnxdeveloper** [comment]
> Almost there! Only formatting issues in two markdown files left: https://github.com/adap/flower/actions/runs/6138546228/job/16655523326?pr=2283#step:6:252

Hi I tried to use format script to reformat these markdown file

File: src/py/flwr/client/rest_client/connection.py [hunk 23]
```diff
@@ -114,17 +120,98 @@ def http_request_response(
...
+        if res.headers["content-type"] != "application/protobuf":
+            log(
+                WARN,
+                "[Node] POST /%s: header `Content-Type` has wrong value",
+                PATH_PULL_TASK_INS,
+            )
+
     def receive() -> Optional[TaskIns]:
         """Receive next task from server."""
         # Serialize ProtoBuf to bytes
```
Comment (danieljanes) [inline src/py/flwr/client/rest_client/connection.py:206]:
```suggestion
```

File: src/py/flwr/client/rest_client/connection.py [hunk 24]
```diff
@@ -114,17 +120,98 @@ def http_request_response(
 
     # Necessary state to link TaskRes to TaskIns
     state: Dict[str, Optional[TaskIns]] = {KEY_TASK_INS: None}
+    # Enable create_node and delete_node to store node
```
Comment (danieljanes) [inline src/py/flwr/client/rest_client/connection.py:123]:
```suggestion

    # Enable create_node and delete_node to store node
```

**Review by danieljanes — CHANGES_REQUESTED** [review]

**Review by danieljanes — APPROVED** [review]
Lgtm, merging! Thanks for the contribution @cnxdeveloper 🎉

**Pull request merged**